Add New Language screen redesign

nayushi opened the pull request:
Light mode:
<img src="https://user-images.githubusercontent.com/69827434/208221264-1ef94c59-1711-42e0-a678-62920da36792.png" width="200" height="400" /> <img src="https://user-images.githubusercontent.com/69827434/208221275-128bb9c0-3c61-4ae9-959f-c6230b57899f.png" width="200" height="400" /> <img src="https://user-images.githubusercontent.com/69827434/208221294-4e511e67-5218-494b-a583-3d7c25a8fab6.png" width="200" height="400" />

Dark mode:
<img src="https://user-images.githubusercontent.com/69827434/208221344-8374ce8c-2572-4db3-bbe3-2b3c2442a7e1.png" width="200" height="400" /> <img src="https://user-images.githubusercontent.com/69827434/208221339-b5497193-d52e-471c-b2c5-85a1a5db2007.png" width="200" height="400" /> <img src="https://user-images.githubusercontent.com/69827434/208221332-a5a5dbfd-4d82-43de-9412-36723d55d3bd.png" width="200" height="400" />

Has blankspace and duplicate validation

misteu commented:
Hello!

I just tested it in the simulator, looks way better already!! ❤️  

However, there are two things:
1. When I am looking at the screen again, what do you think about having a half bottom sheet (or however these are called 😄 ) I think this would look better. Since iOS 15 there is an API for that (`UISheetPresentationController`) so we would not even need some 3rd party for it. 
We can do that in another MR though :) Here are some information about that new API if you did not see already https://sarunw.com/posts/bottom-sheet-in-ios-15-with-uisheetpresentationcontroller/

2. I cannot save a new language. It seems that there was some logic before checking if an array for the languages already exists. You probably did not remove your app when testing that is why it still might have worked for you :)
I don't remember where this logic was before, I'll check but it also works again if you add some method checking and adding an empty array to the UserDefaults, e.g. I just added this for testing purposes in `AppDelegate`:

```swift
  private func createLanguageStoreIfNeeded() {
    if UserDefaults.standard.array(forKey: UserDefaultKeys.languages) == nil {
      UserDefaults.standard.set([], forKey: UserDefaultKeys.languages)
    }
  }
```

You can call that in `AppDelegate.application(_:didFinishLaunchingWithOptions:)` then. Or you do that directly on the `NewLanguageViewController`

nayushi commented:
Hello, the bug is fixed and now for iOS 15+ it shows that way:

<img src="https://user-images.githubusercontent.com/69827434/208268370-fd0486e0-1084-4ddf-b62d-ee9fd6c557fa.png" width="200" height="400" /><img src="https://user-images.githubusercontent.com/69827434/208268373-e56b7b65-c3fe-4ee1-9517-9afd7ce267fc.png" width="200" height="400" />

misteu commented on VocabularyTrainer/Localizable/Localizable.swift:
```diff
@@ -0,0 +1,23 @@
+
+import UIKit
+
+enum Localizable: String {
```
Thanks for thinking about the localized strings as well!

However, I guess this could get a really huge enum at some point if it's global. Or do you plan of having these on a screen basis?

What do you think of changing to MVVM in general starting with this screen? 

Then we could make these Localizable enums an extension to a new `NewLanguageViewModel`. Going a little bit away from the massive view controllers now, will be easier to maintain in the future I think.

A slightly different approach (that I'm currently using in other projects) is making all of the strings static and use an `enum` just for namespacing. Adding the `context` parameter can be valuable for translations/documentation later on. We should consider using it as well. You could use the same approach for any static data like `Images` and  `URLs`.

```swift
extension NewLanguageViewModel {
    enum Strings {
        static let title = NSLocalizedString(
            "title_text",
            value: "The title",
            comment: "Text for the title"
        )
        static let buttonTitle = NSLocalizedString(
            "button_title",
            value: "I am a button",
            comment: "Title of the button"
        )
    }
}
```

in the view controller we could add `typealias`es to make the calling site nicer:

```swift
class NewLanguageViewController: UIViewController {

    typealias Strings = NewLanguageViewModel.Strings

    let viewModel: NewLanguageViewModel
    let button = UIButton()

    init(viewModel: NewLanguageViewModel) {
        self.viewModel = viewModel
    }

    func setUI() {
        button.setTitle(Strings.buttonTitle, for: .normal)
    }
}
```

Let me know what you think :) We can also do the MVVM refactoring in another MR but I think we should do it.

misteu commented on VocabularyTrainer/NewLanguage/NewLanguageViewController.swift:
```diff
@@ -0,0 +1,200 @@
+
+    private lazy var titleLabel: UILabel = .createLabel(font: .systemFont(ofSize: 20, weight: .semibold),
+                                                        text: Localizable.addNewLanguage.localize(),
+                                                        accessibilityTrait: .header)
```
oh nice :) It's great to see A11y right from the start!! ❤️

misteu commented on VocabularyTrainer/NewLanguage/NewLanguageViewController.swift:
```diff
@@ -0,0 +1,200 @@
+        return button
+    }()
+
+    private lazy var titleLabel: UILabel = .createLabel(font: .systemFont(ofSize: 20, weight: .semibold),
```
## AccessibilityPolice 🚓 

When using this initializer you will have a fixed fontsize. We want dynamic type :)

When there are existing iOS fontsizes that work it can be easily replaced with e.g. `.preferredFont(forTextStyle: .body)` (here is a quick overview about all the existing ones: https://www.iosfontsizes.com/ ). If not, you can use `UIFontMetrics` for creating a dynamic one.

We should probably add some helper method/extension to `UIFont` later for quickly creating dynamic type fonts.

I just tried it out this looks good I think:
```swift
    let titleFont = UIFontMetrics(forTextStyle: .body).scaledFont(for: .systemFont(ofSize: 20, weight: .semibold))
    private lazy var titleLabel: UILabel = .createLabel(font: titleFont,
                                                        text: Localizable.addNewLanguage.localize(),
                                                        accessibilityTrait: .header)

    private lazy var languageLabel: UILabel = .createLabel(font: .preferredFont(forTextStyle: .headline),
                                                           text: Localizable.language.localize(),
                                                           accessibilityTrait: .header)
    private lazy var textField: UITextField = {
        let textField = UITextField()
        textField.translatesAutoresizingMaskIntoConstraints = false
        textField.placeholder = Localizable.whichLanguage.localize()
        textField.font = .preferredFont(forTextStyle: .body)

///...
```

For example this it how it would look at size `Accessibility XL` with dynamic type font variants:

![image](https://user-images.githubusercontent.com/35889530/208307568-02e92f43-660b-4093-9ec7-9fa25c1fee10.png)

The default scaling would still look like that:
![image](https://user-images.githubusercontent.com/35889530/208307595-fe54c962-daed-4874-80c3-4040e2aa582b.png)

misteu replied:
ah and I forgot, of course we should do the same for the buttons :)

As you see in my screenshot with increased font size, the button stays the same.

nayushi replied:
<img src="https://user-images.githubusercontent.com/69827434/208325290-a70ca0a6-e534-44ad-a765-1040dbe57f85.png" width="200" height="400" /><img src="https://user-images.githubusercontent.com/69827434/208325303-a5cc8ee8-190d-4ba3-b287-e077a88637e7.png" width="200" height="400" /><img src="https://user-images.githubusercontent.com/69827434/208325321-d6f4ff73-2d4b-43c4-8e32-ef12a634f1de.png" width="200" height="400" />

nayushi commented:
Added: 
- Larger text support
- Added shake movement at Add button for error (duplicate) 
- Haptic feedback for success (adding a new language) and error (duplicate)
- AddSubviews added on UIView extension
- Added focusOn() on UIAccessibility extension

misteu commented on VocabularyTrainer.xcodeproj/project.pbxproj:
```diff
@@ -652,15 +672,15 @@
 				);
 				MARKETING_VERSION = 1.3.0;
-				PRODUCT_BUNDLE_IDENTIFIER = st.mic.VocabularyTrainer;
+				PRODUCT_BUNDLE_IDENTIFIER = st.mic.VocabularyTrainerr;
```
oh what happened here? :D

misteu replied:
oh I see you changed it probably for testing on device, right?

You can keep it like that, I will change it back once preparing the next release.

misteu commented on VocabularyTrainer/NewLanguage/NewLanguageViewController.swift:
```diff
@@ -189,12 +180,38 @@ final class NewLanguageViewController: UIViewController {
         feedbackLabel.isHidden = false
+        addButton.shake()
+        errorHapticFeedback()
+        focusOnErrorMessage()
```
🚀 Very nice!

misteu reviewed (approved):
Great work! Just tested with VoiceOver on the device as well :) 

Also the new Launchscreen really pops with dark mode! ❤️
